pkg/capacity/planner.go

```diff
@@ -0,0 +1,513 @@
+		// HasCapacity checks if the workload could be provisioned with its attached
+		// pool as it is right now.
+		HasCapacity(w workloads.Workloader, seconds uint) (bool, error)
+		// UpdateUsedCapacity for a pool
+		UpdateUsedCapacity(w workloads.Workloader, increase bool) error
+		// PoolByID returns the pool with the given ID
+		PoolByID(id int64) (types.Pool, error)
+		// PoolsForOwner returns all pools for a given owner
+		PoolsForOwner(owner int64) ([]types.Pool, error)
+	}
+
+	// NaivePlanner simply allows all capacity purchases, and allows all workloads
+	// to use a pool, as long as they both have the same owner.
+	NaivePlanner struct {
+		escrow escrow.Escrow
+
+		reserveChan            chan reserveJob
+		allowedChan            chan allowedJob
+		hasCapacityChan        chan hasCapacityJob
+		listChan               chan listPoolJob
+		updateUsedCapacityChan chan updateUsedCapacityJob
+
+		// timer when next pool is empty
+		timer *time.Timer
+
+		db  *mongo.Database
+		ctx context.Context
+	}
+
+	reserveJob struct {
+		reservation  types.Reservation
+		currencies   []string
+		responseChan chan<- reserveResponse
+	}
+
+	reserveResponse struct {
+		info escrowtypes.CustomerCapacityEscrowInformation
+		err  error
+	}
+
+	allowedJob struct {
+		w            workloads.Workloader
+		responseChan chan<- allowedResponse
+	}
+
+	allowedResponse struct {
+		status bool
+		err    error
+	}
+
+	hasCapacityJob struct {
+		w            workloads.Workloader
+		seconds      uint
+		responseChan chan<- hasCapacityResponse
+	}
+
+	hasCapacityResponse struct {
+		status bool
+		err    error
+	}
+
+	listPoolJob struct {
+		id           int64
+		owner        int64
+		responseChan chan<- listPoolResponse
+	}
+
+	listPoolResponse struct {
+		pools []types.Pool
+		err   error
+	}
+
+	updateUsedCapacityJob struct {
+		w            workloads.Workloader
+		used         bool
+		responseChan chan<- updateUsedCapacityResponse
+	}
+
+	updateUsedCapacityResponse struct {
+		err error
+	}
+)
+
+const (
+	unusedPoolExpiration = time.Hour * 24 * 365 * 280
+)
+
+// NewNaivePlanner creates a new NaivePlanner, using the provided escrow and
+// database connection
+func NewNaivePlanner(escrow escrow.Escrow, db *mongo.Database) *NaivePlanner {
+	return &NaivePlanner{
+		escrow:                 escrow,
+		reserveChan:            make(chan reserveJob),
+		allowedChan:            make(chan allowedJob),
+		listChan:               make(chan listPoolJob),
+		hasCapacityChan:        make(chan hasCapacityJob),
+		updateUsedCapacityChan: make(chan updateUsedCapacityJob),
+		db:                     db,
+	}
+}
+
+// Run implements Planner
+func (p *NaivePlanner) Run(ctx context.Context) {
+	p.ctx = ctx
+
+	// first make sure we decomission workloads from expired pools
+	log.Info().Msg("setting up capacity planner expiration timer")
+	if err := p.handlePoolExpiration(true); err != nil {
+		log.Error().Err(err).Msg("failed to expire capacity pools")
+	}
+
+	for {
```

Comment:
As far as I understand this loop and select over many channels is a way to sync access to an internal planner state. Calls to methods basically "blocks" until the response is ready, but I am wondering why a RWMutex was not used instead where methods can just lock it and unlock it on exit. So no channels (and passing channels is needed around)

Reply:
I see this might be done because you can process multiple requests at once, and only wait for the results per request without locking the pool for the process time. In that case ignore my comment

Reply:
Also since this is done in a loop, the planner can't actually process any requests unless if finishes processing the one in hand, so I really don't see why it's done like this

Reply:
I went with channels since I knew there would be a timer channel in the end, and channels can be used as a synchronization primitive. So this keeps the code to only use 1 sync primitive, at the downside of having some boilerplate. Also, we don't really lock anything in the planner, so a mutex does not really convey the proper meaning imo. Also, unlike with  a mutex, this code could be improved in the future to allow multiprocessing if needed.